### ISSUE TITLE
Increase accuracy of receipt-ocr

### ISSUE THREAD

**Issue by starbuck93** [ReceiptManager/receipt-parser-app#8]
I recorded my screen on the Android app to show what's going on. I downloaded the apk from [here](https://apt.izzysoft.de/fdroid/index/apk/org.receipt_manager). Maybe I'm not doing something correctly. It takes nearly 45 seconds to stop loading.

[Google Photos Video link](https://photos.app.goo.gl/Z6WaUTvAjBvyV74ZA)

I opened up Postman to see what the API is up to and here's the response with an uploaded png with the key of "image":

`https://10.0.0.10:8721/api/upload/`
```
{
    "storeName": null,
    "receiptTotal": null,
    "receiptDate": "null",
    "receiptCategory": "grocery"
}
```

The server is running with instructions from the receipt-parser-server readme on Docker with

```
docker pull monolidth/receipt-parser:latest
docker run -p 8721:8721 monolidth/receipt-parser
```

And I had to add a directory in the server Docker container for `/app/data/img` to fix an error.

But further inspection of the container shows that it did OCR the images:

```
root@c1e3b968b0a7:/app# ls data/txt/
receipt_2020-11-30_143843.txt  receipt_2020-11-30_144826.txt  receipt_2020-11-30_151504.txt
receipt_2020-11-30_144445.txt  receipt_2020-11-30_151113.txt  snip_20201130145614.txt
```

As I'm writing my Issue here I'm discovering more.

So I forked the repo and rebuilt the server Docker container with a new config.yml ...

...

It's been a few hours now and I made a few changes to ReceiptManager/receipt-parser now. I found that a regex wasn't matching lines correctly and a `normalize()` function needed another string function to properly split lines and not literally every character.

Anyways... The mobile app still didn't add any receipts as shown in the video. It did, however, fill in the name of my grocery store correctly, which was neat. If I'm missing something please let me know!

**Comment by monolidth** [comment]
Hi @starbuck93. Thank you for _submitting_ your issue. The steps you have done looks correct.

> And I had to add a directory in the server Docker container for /app/data/img to fix an error.

Thanks for catching. I pushed the Docker Image yesterday, see [commit](https://github.com/ReceiptManager/receipt-parser-server/commit/5e6b7f99188649e5c774dcada9b302017ad9d99b) and I did not test this image carefully. I do not use the **Docker Image** at the moment, because it is easier to maintain, using the local version. I will open an ticket on the receipt-parser-server repository.

> But further inspection of the container shows that it did OCR the images

Yeah, the main problem is the camera quality of the mobile phone. You could experiment with the light setting and the zoom to increase accuracy.

> It's been a few hours now and I made a few changes to `ReceiptManager/receipt-parser` now. I found that a regex wasn't matching lines correctly and a normalize() function needed another string function to properly split lines and not literally every character.

It would be a pleasure if you could submit an pull request.  I have to take a look at the the normalize() function tomorrow. Since it is 1AM and I'm very tired.

> Anyways... The mobile app still didn't add any receipts as shown in the video. It did, however, fill in the name of my grocery store correctly, which was neat. If I'm missing something please let me know!

You could take a look at tesseract output located at: `data/txt`. Maybe you could add a new `total` identifier. Sometimes a digit in the receipt date flip, which causes an invalid date. The parser handle this with `null`

I hope I could clarify this and I would be happy to see enhancements :)

Regards,
William

**Comment by monolidth** [comment]
See: https://github.com/ReceiptManager/receipt-parser-server/issues/16

**Comment by starbuck93** [comment]
I plan on working on this further as I've been looking for a nifty receipts "solution" for a few months now. I have a large Paperless database currently but they're all just tagged as `receipt` with the vendor and that's not particularly helpful for receipt tracking. I wish I knew more about Android apps so I could further debug what's happening as regards to this specific repo and issue. Perhaps I'll take a stab at running a dev build of that as well.

Also, I don't know if this issue,  #8, is so much about the "Inaccuracy of OCR" versus I have no idea why no receipts didn't pop up in the "Receipt Manager" tab in the Android app. Based on my experience with tesseract via Paperless, the OCR recognition `txt ` is fine.

**Comment by monolidth** [comment]
> I plan on working on this further as I've been looking for a nifty receipts "solution" for a few months now.
Yes, that is the main reason why I started the project. Most of the current solution does not work or are very limited.

> Also, I don't know if this issue, #8, is so much about the "Inaccuracy of OCR" versus I have no idea why no receipts didn't pop up in the "Receipt Manager" tab in the Android 
No, this has nothing todo with the android app. The android app does not parse the receipt. It only take the submited json receipt and insert this into the SQL database. If you want to investigate the problem more. Do the following

1. Clone this repository: https://github.com/ReceiptManager/receipt-parser
2. Install the image magic lib dependency
3. Move your receipt photo to `data/img`
4. Run `make install && make run`

>  Based on my experience with tesseract via Paperless, the OCR recognition txt is fine. 
You can post the tesseract output here. This would help a lot :)

**Comment by starbuck93** [comment]
Here's some [Tesseract txt output](https://hastebin.com/jesoxifewa.properties) with my [config.yml](https://hastebin.com/aduzeqakaq.properties) and the [test image](https://i.imgur.com/ANfYFBa.jpg) I uploaded with Postman. (since this was a scanned image at 300 dpi it won't reflect the poor quality of mobile cameras)

The json result: https://hastebin.com/hiduqayiya.json

And some server output: https://hastebin.com/ugiqadeqod.yaml

Attempting to add a receipt via mobile still did not work. Should the app still let you fill in the details (in my case the date was missing) and still add the receipt?

Here's what that looks like: [server output](https://hastebin.com/zowuputana.yaml) - [tesseract output](https://hastebin.com/evoyoholab.nginx) - [uploaded result](https://i.imgur.com/PLKG82y.png)

**Comment by starbuck93** [comment]
> purple date button

That's the solution right there! I was tapping on the text field and manually typing in a date. As soon as I used the datepicker instead of the text field it added the receipt to the middle tab.

**Comment by monolidth** [comment]
I fixed all dates error in the last view commits. I check now the current local of the device and select the corresponding date format string.

Now, you can insert the date into the text field. The bug was, that the input was never read. I simplify the method and hope that everything work as expected.

Regards,
deltax

**Comment by monolidth** [comment]
Is this issue resolved in ReceiptManager 1.0.4+21?
